File: backend/internal/marble/marble.go
package marble

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

const baseURL = "https://api.marblecms.com"

var httpClient = &http.Client{Timeout: 10 * time.Second}

type Author struct {
	Name  string `json:"name"`
	Image string `json:"image"`
}

type Tag struct {
	Name string `json:"name"`
	Slug string `json:"slug"`
}

type Category struct {
	Name string `json:"name"`
	Slug string `json:"slug"`
}

type Post struct {
	ID          string   `json:"id"`
	Slug        string   `json:"slug"`
	Title       string   `json:"title"`
	Description string   `json:"description"`
	Content     string   `json:"content"`
	CoverImage  string   `json:"coverImage"`
	PublishedAt string   `json:"publishedAt"`
	UpdatedAt   string   `json:"updatedAt"`
	Authors     []Author `json:"authors"`
	Category    Category `json:"category"`
	Tags        []Tag    `json:"tags"`
}

type listResponse struct {
	Posts []Post `json:"posts"`
}

type getResponse struct {
	Post *Post `json:"post"`
}

func apiGet(path string) ([]byte, error) {
	key := os.Getenv("MARBLE_API_KEY")
	if key == "" {
		return nil, fmt.Errorf("MARBLE_API_KEY not set")
	}

	req, err := http.NewRequest(http.MethodGet, baseURL+path, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", key)
	req.Header.Set("Accept", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("marble api %s: %s", resp.Status, strings.TrimSpace(string(body)))
	}
	return body, nil
}

func ListPosts() []Post {
	body, err := apiGet("/v1/posts")
	if err != nil {
		log.Printf("marble ListPosts: %v", err)
		return nil
	}

	var parsed listResponse
	if err := json.Unmarshal(body, &parsed); err != nil {
		var posts []Post
		if err2 := json.Unmarshal(body, &posts); err2 != nil {
			log.Printf("marble ListPosts: decode: %v", err2)
			return nil
		}
		return posts
	}
	return parsed.Posts
}

func GetPost(slug string) *Post {
	body, err := apiGet("/v1/posts/" + slug)
	if err != nil {
		log.Printf("marble GetPost %q: %v", slug, err)
		return nil
	}

	var wrapped getResponse
	if err := json.Unmarshal(body, &wrapped); err == nil && wrapped.Post != nil {
		return wrapped.Post
	}

	var post Post
	if err := json.Unmarshal(body, &post); err != nil || post.Slug == "" {
		log.Printf("marble GetPost %q: decode failed or empty slug", slug)
		return nil
	}
	return &post
}

func VerifySignature(secret, signatureHeader, bodyText string) bool {
	expectedHex := strings.TrimPrefix(signatureHeader, "sha256=")
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(bodyText))
	return expectedHex == hex.EncodeToString(mac.Sum(nil))
}

File: src/lib/marble.ts
export type Author = {
  name: string;
  image: string;
};

export type Tag = {
  name: string;
  slug: string;
};

export type Post = {
  id: string;
  slug: string;
  title: string;
  description: string;
  content: string;
  coverImage: string;
  publishedAt: string;
  updatedAt: string;
  authors: Author[];
  tags: Tag[];
};

const baseUrl = "https://api.marblecms.com";

async function apiGet(path: string): Promise<Response> {
  const key = import.meta.env.MARBLE_API_KEY;
  if (!key) {
    throw new Error("MARBLE_API_KEY not set");
  }

  return fetch(`${baseUrl}${path}`, {
    headers: {
      Authorization: key,
      Accept: "application/json",
    },
  });
}

export async function listPosts(): Promise<Post[]> {
  try {
    const response = await apiGet("/v1/posts");
    if (!response.ok) {
      console.error("marble ListPosts:", response.status, await response.text());
      return [];
    }

    const parsed = (await response.json()) as { posts: Post[] };
    return parsed.posts ?? [];
  } catch (error) {
    console.error("marble ListPosts:", error);
    return [];
  }
}

export async function getPost(slug: string): Promise<Post | null> {
  try {
    const response = await apiGet(`/v1/posts/${slug}`);
    if (!response.ok) {
      console.error(
        "marble GetPost:",
        slug,
        response.status,
        await response.text(),
      );
      return null;
    }

    const parsed = (await response.json()) as { post: Post | null };
    return parsed.post ?? null;
  } catch (error) {
    console.error("marble GetPost:", slug, error);
    return null;
  }
}

export function verifySignature(
  secret: string,
  signatureHeader: string,
  bodyText: string,
): boolean {
  const expected = signatureHeader.replace(/^sha256=/, "");
  const encoder = new TextEncoder();
  const keyData = encoder.encode(secret);
  const messageData = encoder.encode(bodyText);

  return crypto.subtle
    .importKey("raw", keyData, { name: "HMAC", hash: "SHA-256" }, false, [
      "sign",
    ])
    .then((key) => crypto.subtle.sign("HMAC", key, messageData))
    .then((signature) => {
      const hash = Array.from(new Uint8Array(signature))
        .map((byte) => byte.toString(16).padStart(2, "0"))
        .join("");
      return hash === expected;
    })
    .catch(() => false);
}

export function formatPublished(raw: string): string {
  if (!raw) {
    return "";
  }

  const date = new Date(raw);
  if (Number.isNaN(date.getTime())) {
    return raw;
  }

  return date.toLocaleDateString("en-US", {
    month: "long",
    day: "numeric",
    year: "numeric",
  });
}

export function getAuthor(post: Post): { name: string; image: string } {
  const author = post.authors[0];
  return {
    name: author?.name || "Valtteri Savonen",
    image: author?.image || "",
  };
}
